vnext/Shared/JSI/NapiJsiRuntime.cpp

```diff
@@ -0,0 +1,1921 @@
+    NapiRefHolder Symbol;
+    NapiRefHolder byteLength;
+    NapiRefHolder configurable;
+    NapiRefHolder enumerable;
+    NapiRefHolder get;
+    NapiRefHolder getOwnPropertyDescriptor;
+    NapiRefHolder hostFunctionSymbol;
+    NapiRefHolder hostObjectSymbol;
+    NapiRefHolder length;
+    NapiRefHolder message;
+    NapiRefHolder ownKeys;
+    NapiRefHolder propertyIsEnumerable;
+    NapiRefHolder prototype;
+    NapiRefHolder set;
+    NapiRefHolder toString;
+    NapiRefHolder value;
+    NapiRefHolder writable;
+  } m_propertyId;
+
+  // Cache of commonly used values.
+  struct CachedValue final {
+    NapiRefHolder Error;
+    NapiRefHolder Global;
+    NapiRefHolder False;
+    NapiRefHolder HostObjectProxyHandler;
+    NapiRefHolder Null;
+    NapiRefHolder ProxyConstructor;
+    NapiRefHolder SymbolToString;
+    NapiRefHolder True;
+    NapiRefHolder Undefined;
+  } m_value;
+
+  bool m_pendingJSError{false};
+};
+} // namespace
+
+} // namespace Microsoft::JSI
+
+#pragma endregion Declarations
+
+namespace Microsoft::JSI {
+
+namespace {
+constexpr char s_description[] = "NapiJsiRuntime";
+
+#pragma region NapiJsiRuntime
+
+NapiJsiRuntime::NapiJsiRuntime(napi_env env) noexcept : m_env{env} {
+  EnvScope scope{m_env};
+  m_propertyId.Error = NapiRefHolder{this, GetPropertyIdFromName("Error"sv)};
+  m_propertyId.Object = NapiRefHolder{this, GetPropertyIdFromName("Object"sv)};
+  m_propertyId.Proxy = NapiRefHolder{this, GetPropertyIdFromName("Proxy"sv)};
+  m_propertyId.Symbol = NapiRefHolder{this, GetPropertyIdFromName("Symbol"sv)};
+  m_propertyId.byteLength = NapiRefHolder{this, GetPropertyIdFromName("byteLength"sv)};
+  m_propertyId.configurable = NapiRefHolder{this, GetPropertyIdFromName("configurable"sv)};
+  m_propertyId.enumerable = NapiRefHolder{this, GetPropertyIdFromName("enumerable"sv)};
+  m_propertyId.get = NapiRefHolder{this, GetPropertyIdFromName("get"sv)};
+  m_propertyId.getOwnPropertyDescriptor = NapiRefHolder{this, GetPropertyIdFromName("getOwnPropertyDescriptor"sv)};
+  m_propertyId.hostFunctionSymbol = NapiRefHolder{this, CreateSymbol("hostFunctionSymbol"sv)};
+  m_propertyId.hostObjectSymbol = NapiRefHolder{this, CreateSymbol("hostObjectSymbol"sv)};
+  m_propertyId.length = NapiRefHolder{this, GetPropertyIdFromName("length"sv)};
+  m_propertyId.message = NapiRefHolder{this, GetPropertyIdFromName("message"sv)};
+  m_propertyId.ownKeys = NapiRefHolder{this, GetPropertyIdFromName("ownKeys"sv)};
+  m_propertyId.propertyIsEnumerable = NapiRefHolder{this, GetPropertyIdFromName("propertyIsEnumerable"sv)};
+  m_propertyId.prototype = NapiRefHolder{this, GetPropertyIdFromName("prototype"sv)};
+  m_propertyId.set = NapiRefHolder{this, GetPropertyIdFromName("set"sv)};
+  m_propertyId.toString = NapiRefHolder{this, GetPropertyIdFromName("toString"sv)};
+  m_propertyId.value = NapiRefHolder{this, GetPropertyIdFromName("value"sv)};
+  m_propertyId.writable = NapiRefHolder{this, GetPropertyIdFromName("writable"sv)};
```

Comment:
all these CreateSymbol (and probably other property ID from name) will result in strings that will be duplicated across every single instance. Make the strings class statics const to avoid the string duplication since the utf8 string will always be the same and not memoized this way

Reply:
There will be no C++ constant string duplications, but the NapiRefHolder cannot be global, it must be per the V8 engine runtime instance.